components/list/src/provider/serde_dfa.rs

```diff
@@ -52,58 +63,57 @@ impl serde::Serialize for StringMatcher<'_> {
                 .unwrap_or_else(|| {
                     use serde::ser::Error;
                     Err(S::Error::custom(
-                        "cannot serialize a deserialized bincode StringMatcher to JSON",
+                        "cannot serialize a deserialized bincode SerdeDFA to JSON",
                     ))
                 })
         } else {
-            self.dfa_bytes.serialize(serializer)
+            self.deref().to_bytes_little_endian().serialize(serializer)
         }
     }
 }
 
 #[cfg(feature = "serde")]
-impl<'de: 'data, 'data> serde::Deserialize<'de> for StringMatcher<'data> {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+impl<'data> SerdeDFA<'data> {
+    /// Deserializes to `Option<Self>`. Will return `None` for non-human-readable serialization
+    /// formats on big-endian systems, as `regex_automata` serialization is endian-sensitive.
+    pub fn maybe_deserialize<'de: 'data, D>(deserializer: D) -> Result<Option<Self>, D::Error>
     where
         D: serde::de::Deserializer<'de>,
     {
         use icu_provider::serde::borrow_de_utils::CowBytesWrap;
+        use serde::Deserialize;
 
         #[cfg(feature = "serde_human")]
         if deserializer.is_human_readable() {
             #[cfg(not(feature = "std"))]
             use alloc::string::ToString;
             use serde::de::Error;
-            return StringMatcher::new(Cow::<str>::deserialize(deserializer)?)
+            return SerdeDFA::new(Cow::<str>::deserialize(deserializer)?)
+                .map(Some)
                 .map_err(|e| D::Error::custom(e.to_string()));
         }
 
+        let dfa_bytes = <CowBytesWrap<'de>>::deserialize(deserializer)?.0;
+
         if cfg!(target_endian = "big") {
-            // TODO: Convert LE to BE. For now we just behave like the
-            // accept-nothing DFA on BE systems.
-            return Ok(StringMatcher {
-                dfa_bytes: Cow::Borrowed(&[]),
-                pattern: None,
-            });
+            return Ok(None);
         }
 
-        let dfa_bytes = <CowBytesWrap<'de>>::deserialize(deserializer)?.0;
-
         // Verify safety invariant
         DFA::from_bytes(&dfa_bytes).map_err(|e| {
```

Comment:
Observation: You could envision a BE system with baked data providing a data service to LE clients. In this case, the DFA validator will run with improper data. I assume we will get the expected behavior of these data structs simply failing to deserialize.

Reply:
> BE system with baked data providing a data service to LE clients

How? BE and LE systems cannot share memory. They will have to go through serde.

Reply:
(and going through serde serializes as LE, not native, so that works from BE systems)